extensions/markdown-language-features/src/features/smartSelect.ts

```diff
@@ -0,0 +1,238 @@
+			}
+		}
+		if (currentRange) {
+			return currentRange;
+		} else {
+			return parentRange;
+		}
+	}
+
+	private async getHeaderSelectionRange(document: vscode.TextDocument, position: vscode.Position): Promise<vscode.SelectionRange | undefined> {
+		const tocProvider = new TableOfContentsProvider(this.engine, document);
+		const toc = await tocProvider.getToc();
+
+		let headerInfo = getHeadersForPosition(toc, position);
+
+		let headers = headerInfo.headers;
+
+		let parentRange: vscode.SelectionRange | undefined;
+		let currentRange: vscode.SelectionRange | undefined;
+
+		for (let i = 0; i < headers.length; i++) {
+			currentRange = createHeaderRange(i === headers.length - 1, headerInfo.headerOnThisLine, headers[i], parentRange, getFirstChildHeader(document, headers[i], toc));
+			if (currentRange && currentRange.parent) {
+				parentRange = currentRange;
+			}
+		}
+		return currentRange;
+	}
+}
+
+function getFirstChildHeader(document: vscode.TextDocument, header?: TocEntry, toc?: TocEntry[]): vscode.Position | undefined {
+	let childRange: vscode.Position | undefined;
+	if (header && toc) {
+		let children = toc.filter(t => header.location.range.contains(t.location.range) && t.location.range.start.line > header.location.range.start.line).sort((t1, t2) => t1.line - t2.line);
+		if (children.length > 0) {
+			childRange = children[0].location.range.start;
+			let lineText = document.lineAt(childRange.line - 1).text;
+			return childRange ? childRange.translate(-1, lineText.length) : undefined;
+		}
+	}
+	return undefined;
+}
+
+function getTokensForPosition(tokens: Token[], position: vscode.Position): Token[] {
+	let enclosingTokens = tokens.filter(token => token.map && (token.map[0] <= position.line && token.map[1] > position.line) && isBlockElement(token));
+
+	if (enclosingTokens.length === 0) {
+		return [];
+	}
+
+	let sortedTokens = enclosingTokens.sort((token1, token2) => (token2.map[1] - token2.map[0]) - (token1.map[1] - token1.map[0]));
+	return sortedTokens;
+}
+
+function getHeadersForPosition(toc: TocEntry[], position: vscode.Position): { headers: TocEntry[], headerOnThisLine: boolean } {
+	let enclosingHeaders = toc.filter(header => header.location.range.start.line <= position.line && header.location.range.end.line >= position.line);
+	let sortedHeaders = enclosingHeaders.sort((header1, header2) => (header1.line - position.line) - (header2.line - position.line));
+	let onThisLine = toc.find(header => header.line === position.line) !== undefined;
+	return {
+		headers: sortedHeaders,
+		headerOnThisLine: onThisLine
+	};
+}
+
+function isBlockElement(token: Token): boolean {
+	return !['list_item_close', 'paragraph_close', 'bullet_list_close', 'inline', 'heading_close', 'heading_open'].includes(token.type);
+}
+
+function createHeaderRange(isClosestHeaderToPosition: boolean, onHeaderLine: boolean, header?: TocEntry, parent?: vscode.SelectionRange, childStart?: vscode.Position): vscode.SelectionRange | undefined {
```

Comment:
[nit] Check if `header` needs to be nullable. If not, you could also update `createHeaderRange` to never return `undefined`